File: library_managment/library_managment/report/report_for_article/report_for_article.py
import frappe

# def execute(filters=None):
    # frappe.msgprint(str(filters))
    # if not filters:
    #     filters = {}

    # # Columns define kar
    # columns = [
    #     {"label": "Article Name", "fieldname": "article", "fieldtype": "Data", "width": 200},
    #     {"label": "Author", "fieldname": "author", "fieldtype": "Data", "width": 150},
    #     {"label": "Publisher", "fieldname": "publisher", "fieldtype": "Data", "width": 120},
    #     {"label": "status", "fieldname": "status", "fieldtype": "Data", "width": 120},
    # ]

    # # Filter ke hisaab se conditions banao
    # conditions = []

    # if filters.get("author"):
    #     conditions.append("LOWER(author) LIKE LOWER(%(author)s)")
    #     filters["author"] = f"%{filters['author']}%"

    # if filters.get("article"):
    #     conditions.append("LOWER(article) LIKE LOWER(%(article)s)")
    #     filters["article"] = f"%{filters['article']}%"

    # condition_str = ""
    # if conditions:
    #     condition_str = "WHERE " + " AND ".join(conditions)

    # # SQL query likh
    # query = f"""
    #     SELECT
    #         article,
    #         author,
    #         publisher
    #     FROM `tabArticle`
    #     {condition_str}
    #     ORDER BY modified DESC
    # """

    # data = frappe.db.sql(query, filters, as_dict=True)

    # return columns, data


# import frappe

def execute(filters=None):
    columns = [
        {"fieldname": "article", "label": "Article", "fieldtype": "Data"},
        {"fieldname": "author", "label": "Author", "fieldtype": "Data"},
        {"fieldname": "publisher", "label": "Publisher", "fieldtype": "Data"}
    ]

    data = frappe.db.sql("""
        SELECT
            article,
            author,
            publisher
        FROM `tabArticle`
        ORDER BY modified DESC
    """, as_dict=True)

    # Chart data
    chart = {
        "data": {
            "labels": [d.article for d in data],
            "datasets": [
                {
                    "name": "Articles Count",
                    "values": [1 for d in data]   # simple value for each article
                }
            ]
        },
        "type": "bar",   # can be 'line', 'bar', 'pie'
        "height": 300
    }

    return columns, data, None, chart

File: library_managment/library_managment/report/report_for_article/report_for_article.js
// Copyright (c) 2025, codeWithSwapnil and contributors
// For license information, please see license.txt

frappe.query_reports["report for article"] = {
	"filters": [

			{
			"fieldname":"author",
			"fieldtype": "Data",
			"label": "author"
			},
			{
			"fieldname":"article",
			"fieldtype": "Link",
			"label": "articles",
			"options":"Article"
			},
			// {
			// 	"fieldname":"from_date",
			// 	"fieldtype":"Date",
			// 	"label":"from date",
			// 	"default":dateutil.year_start()
			// }

	],
	//  "onload": function (report) {
	// 	// Chart create karne ke liye initial config
	// 	report.page.add_inner_button(__('Show Chart'), function () {
	// 		let data = report.data;

	// 		// Count of articles by author
	// 		let chart_data = {};
	// 		data.forEach(d => {
	// 			if (chart_data[d.author]) {
	// 				chart_data[d.author] += 1;
	// 			} else {
	// 				chart_data[d.author] = 1;
	// 			}
	// 		});

	// 		const labels = Object.keys(chart_data);
	// 		const values = Object.values(chart_data);

	// 		frappe.show_chart({
	// 			data: {
	// 				labels: labels,
	// 				datasets: [
	// 					{
	// 						name: "Articles Count",
	// 						values: values
	// 					}
	// 				]
	// 			},
	// 			type: 'pie',  // 'bar' for bar chart
	// 			title: 'Articles by Author'
	// 		});
	// 	});
	// }
};
